### PR TITLE
Started to revise the pep

### PR DESCRIPTION
**Pull request by TobiasWrigstad**

### CODE REVIEW

File: pep-9999.rst [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+  }  
+  ... // code that performs the write
+
+
+Writes are common in the CPython code base and the writes lack a
+common "code path" that they pass. To this end, the PEP requires a
+``Py_CHECKWRITE`` call to be inserted and there are several places
+in the CPython code base that are changed as a consequence of this
+PEP. So far we have identified around 70 writes spread across a
+dozen files. **TODO: update this number**
```
Comment (TobiasWrigstad) [inline pep-9999.rst:189]: @matajoh Can you update this paragraph?

File: pep-9999.rst [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+failure to freeze an object graph. Consider the object graph
+``o1 --> o2 --> o3`` where ``o1`` and ``o3`` can be frozen, but
+``o2`` cannot. What are the possible behaviours of ``freeze(o1)``?
+
+1. Freeze fails completely. In the strict interpretation of deep
+   immutability, freezing ``o1`` is not possible because ``o1``
+   contains a reference to an un-freezable object ``o2``. In this
+   scenario, the object graph ``o1 --> o2 --> o3`` remains mutable
+   and ``freeze(o1)`` raises an exception when the object graph
+   traversal encounters ``o2``.
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:263]: This is a great point! I am not sure that is as user-friendly, but it would promote more objects becoming immutable. Should we move this discussion to Teams?

File: pep-9999.rst [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+To opt-in support for immutability, C extensions with functions
+that directly write to data which can be made immutable should add
+the ``Py_CHECKWRITE`` macro shown above on all paths in the code
+that lead to writes to that data. Notably, if C extensions manage
+their data through Python objects, no changes are needed.
+
+Python objects may define a ``__freeze__`` method which will be
+called **after** an object has been made immutable. This hook can
+be used to freeze or otherwise manage any other state on the side
+that is introduced through a C-extension.
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:314]: Sorry, I don't understand what you mean.

File: pep-9999.rst [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+
+      def foo():
+          return x
+
+      freeze(foo)
+      ... # some code, e.g. pass foo to another thread
+      x = 1
+      foo()
+
+  example1()
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:532]: Just so that I am clear, will freezing `foo()` behave differently depending on whether `foo()` is in the global scope or a nested function? That does not feel great to me!

File: pep-9999.rst [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+.. code-block:: python
+
+  from freeze import freeze
+
+  def example2():
+      x = 0
+      def foo(a = False):
+          if a:
+              a = a + 1
+              return a
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:570]: The idea is to illustrate that mutating local variables work — I'd like to keep this line. Maybe we can just add a comment saying `# updating local variables work, even in a frozen function`. Would that work?

File: pep-9999.rst [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+    2. While more permissive, the model suffers from non-determinism.
+       Data races can be hidden in corner cases that require complex
+       logic and/or temporal interactions which can be hard to
+       test and reproduce.
+
+
+Open Issues
+===========
+
+1. How does deep freezing interact with weak references?
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:664]: Will you put on your TODO to write a short section on this?

File: pep-9999.rst [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+---------------------
+
+While useful on their own, all the changes above are building
+blocks of Data-Race Free Python. Data-Race Free Python will
+borrow concepts from ownership (namely region-based ownership,
+see e.g. `Cyclone <https://cyclone.thelanguage.org/>`_) to make Python programs data-race free
+by construction. Which will permit multiple subinterpreters to
+share *mutable* state, although only one subinterpreter at a time
+will be able to access (read or write) to that state. In theory,
+this work could also be authored on-top of free-theaded Python (PEP 703).
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:767]: Yes, great point! Do we have a clear understanding of what we can put up as a preprint?

File: pep-9999.rst [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+
+Reference Implementation
+========================
+
+`Available here <https://github.com/mjp41/cpython/pull/51>`_.
+
+Rebasing on Python 3.14
+=======================
+
+**TODO**
```
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:789]: Good point. A strawman would be to freeze only trees…

File: pep-9999.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply pep-9999.rst:666]: So I just hit a bug in the SCC prototype related to this.  Should a weak reference be converted to an effectively strong reference if it is frozen?

Consider:
```mermaid
stateDiagram
  r1 --> a
  a  --w--> b
  r2 --> b
```
If we `freeze(r1)`, and then `del(r2)`.  Should `a` remain reachable?

Consider:
```mermaid
stateDiagram
  r1 --> a
  a  --w--> b
  b  --> a
```

If we `freeze(a)`, should `a` and `b` be in the same SCC.

File: pep-9999.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply pep-9999.rst:534]: That is how Python works.  Global scoped functions capture globals by name, where as inside a function scope, they are boxed into PyCells.

File: pep-9999.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply pep-9999.rst:572]: Sure.  It needs explaining as without explanation it looks strange to me.

File: pep-9999.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply pep-9999.rst:768]: I think the version we submitted as Camera Ready.

File: pep-9999.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:534]: I understand that. But I am just second guessing our function freezing design that exposes this to users.

File: pep-9999.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:572]: Yes, great.

[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:572]:
```suggestion
          if a:
              a = a + 1 # Note: updating local variables work, even in a frozen function
              return a
```

File: pep-9999.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:768]: I'll upload that and link

File: pep-9999.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:790]: I was thinking that if the RC=1 (== tree) then there won't be any mutation under the hood.

File: pep-9999.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:666]: This is a super good point! I'll need to think about this.

File: pep-9999.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply pep-9999.rst:790]: Ah, okay, that would work, but is pretty restrictive

File: pep-9999.rst [hunk 9]
```diff
@@ -0,0 +1,850 @@
...
+-----------------------
+
+Python programs frequently manipulate large, interconnected data
+structures such as dictionaries, lists, and user-defined objects.
+Unintentional mutations can introduce subtle and
+difficult-to-debug errors. By allowing developers to explicitly
+freeze objects and their transitive dependencies, Python can
+provide stronger correctness guarantees for data processing
+pipelines, functional programming paradigms, and API boundaries
+where immutability is beneficial.
```
Comment (xFrednet) [inline pep-9999.rst:83]: Should we maybe mention caching here? So the fact that objects returned from a cache are usually mutable which can lead to interesting results if they changed?

File: pep-9999.rst [hunk 10]
```diff
@@ -0,0 +1,850 @@
+PEP: 9999
+Title: Deep Immutability in Python
+Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>
+Sponsor: TBD
```
Comment (TobiasWrigstad) [inline pep-9999.rst:4]: Who should be our sponsor? Does that happen *now* or later?

File: pep-9999.rst [hunk 11]
```diff
@@ -0,0 +1,850 @@
...
+Discussions-To: TBD
```
Comment (TobiasWrigstad) [inline pep-9999.rst:5]: Do we know how to open a discussion on discord?

File: pep-9999.rst [hunk 12]
```diff
@@ -0,0 +1,850 @@
...
+Resolution:
+
+
+Abstract
+========
+
+This PEP proposes adding a mechanism for deep immutability to
+Python. The mechanism requires some changes to the core language,
+but user-facing functions are delivered in a module called
+``freeze``. This module provides two functions:
```
Comment (TobiasWrigstad) [inline pep-9999.rst:21]: Module called `immutable`

File: pep-9999.rst [hunk 13]
```diff
@@ -0,0 +1,850 @@
...
+Making an object *deeply* immutable recursively renders the object
+*and all objects it references* immutable. (Just
+making the first object immutable is called *shallow*
+immutability.)
+
+Deep immutability provides strong guarantees against
+unintended modifications, thereby improving correctness, security, and
+parallel execution safety.
+
+Immutable objects are not immortal -- their reference counts
```
Comment (TobiasWrigstad) [inline pep-9999.rst:35]: not immortal by default // orthogonal

File: pep-9999.rst [hunk 14]
```diff
@@ -0,0 +1,850 @@
...
+      def dec(self):
+          self.value -= 1
+      def get(self):
+          return self.value
+
+  c = Counter(0)
+  c.get() # returns 0 
+  freeze(c) # (*) -- this locks the value of the counter to 0
+  ... 
+  Counter.get = lambda self: 42 # OK in CPython 3.12, throws exception with this PEP
```
Comment (TobiasWrigstad) [inline pep-9999.rst:483]: Clarify why

File: pep-9999.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst:540]: Following discussion with @mjp41 -- we keep the above and aim to treat freezing functions uniformly regardless of the nature of the enclosing scope. I will also add a note saying that the prototype does not yet do this.

File: pep-9999.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (TobiasWrigstad) [reply pep-9999.rst]: note: we discussed locking and backing out on races (including freeze-freeze races…)

File: pep-9999.rst [hunk 15]
```diff
@@ -0,0 +1,868 @@
...
+  common cell object.
+- The ``is_freezable(obj)`` function is not yet implemented.
+- The ``NotFreezable`` type is currently freezable (but inheriting
+  from it stops instances of the inheriting class from being frozen). 
+
+
+Rebasing on Python 3.14
+=======================
+
+**TODO MJP**
```
Comment (mjp41) [inline pep-9999.rst:847]:
```suggestion
We have found two areas that need to be addressed to integrate this work with "free-threaded Python": data-representation and data-races during freeze.

Data-representation for immutability
------------------------------------

With free-threaded Python the representation of the reference count has been changed.  We could either borrow a bit to represent if an object is frozen, or alternatively, we could use the new `ob_tid` field to have a special value for immutable state.  Using `ob_tid` would allow for standard mutable thread local objects to remain the fast path, and is our preferred alternative.

The extensions use use SCC calculations to detect cycles in immutable graphs, would require additional state.  Repurposing `ob_tid` and `ob_ref_shared` would allow sufficient space for the necessary calculation.

Data-races during freeze
------------------------

We consider the following races

- Freezing some objects concurrently with another thread checking if a graph is immutable.
- Freezing some objects concurrently with another thread mutating those objects.
- Freezing some objects concurrently with another thread freezing those objects.

To address the first race, we need to consider strictness of deep immutability.  We need to ensure that querying an object graph for immutability only says yes if it is deeply immutable. This requires a two step immutable state: immutable but not strict, and then immutable and strict. On a DFS traversal of the object graph items are marked as immutable but not strict on the pre-order step, and then immutable and strict on the post-order step.  To query if a graph is immutable, we will require the "immutable and strict" state.

Handling mutation during freeze can use the mutex added by free-threading. There are some cases where mutation does not require the acquisition of a mutex, which would no longer allowed with this feature.  Freezing would be required to lock the object, marks it as immutable, release the lock, and then read all its fields.

The final case is the most complex detecting parallel freezing of an object graph.  We will consider this an error.  This error can be detected as follows.  If we encounter an object that is "immutable but not strict", then this should be on the path to the current object from the starting point of the freeze.  If this is not the case, then we must be observing another thread freezing an object graph.  The algorithm should back out the pending aspects of freeze, and raise an exception to the user.  This can naturally be integrated with the SCC algorithm. 
```

File: pep-9999.rst [hunk 16]
```diff
@@ -0,0 +1,868 @@
...
+Abstract
+========
+
+This PEP proposes adding a mechanism for deep immutability to
+Python. The mechanism requires some changes to the core language,
+but user-facing functions are delivered in a module called
+``immutable``. This module provides the following functions and types:
+
+1. The function ``freeze(obj)`` -- turns ``obj`` deeply immutable
+2. The function ``is_frozen(obj)`` -- returns ``True`` if ``obj`` is immutable
```
Comment (matajoh) [inline pep-9999.rst:24]: `isfrozen`

File: pep-9999.rst [hunk 17]
```diff
@@ -0,0 +1,868 @@
...
+3. The function ``is_freezable(obj)`` -- returns ``True`` if all objects reachable from ``obj`` can be frozen
```
Comment (matajoh) [inline pep-9999.rst:25]: `isfreezable`

File: pep-9999.rst [hunk 18]
```diff
@@ -0,0 +1,868 @@
...
+  f.f = g
+  g.f = h
+  h.f = g # cycles are OK!
+  del g # Remove local ref to g, so g's RC = 1
+  del h # Remove local reg to h, so h's RC = 1
+
+  g.x = "African Swallow" # OK
+  freeze(f) # Makes, f, g and h immutable
+  g.x = "European Swallow" # Throws an exception "g is immutable"
+  is_frozen(h) # returns True
```
Comment (matajoh) [inline pep-9999.rst:69]: `isfrozen`

File: pep-9999.rst [hunk 19]
```diff
@@ -0,0 +1,868 @@
...
+
+This proposal intends to be fully backward compatible, as no existing Python
+code will be affected unless it explicitly calls ``freeze(obj)``.
+Frozen objects will raise errors only when mutation is attempted.
+
+
+Opt-In vs. Opt-Out
+------------------
+
+All pure Python objects can be frozen. However, for types which
```
Comment (matajoh) [inline pep-9999.rst:226]: suggested edit: All pure Python objects can be implicitly frozen, provided all their members and their base classes can be frozen.

File: pep-9999.rst [hunk 20]
```diff
@@ -0,0 +1,868 @@
...
+immutability requires some work on both exposing objects to
+freezing, and to enforce immutability in mutating C-functions.
+
+From a backwards compatibility perspective, an opt-in model keeps
+things simple: all existing code keeps working, and only code that
+wishes to support immutability needs updating. The downside of the
+opt-in model is that a large part of all Python libraries cannot
+be (even nominally) frozen.
+
+This PEP proposes to make support for immutability in C extensions
```
Comment (matajoh) [inline pep-9999.rst:237]: add: and Python wrappers of classes which would otherwise not be freezable

File: pep-9999.rst [hunk 21]
```diff
@@ -0,0 +1,868 @@
...
+
+Due to the opt-in decision, there are no *obligations* for C
+extensions that does not want to add support for immutability. 
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box" as long as they use Python standard types to store
+data and uses the built-in functions of these types to modify the data.
+
+To make its instances freezable, types using C extensions must
```
Comment (matajoh) [inline pep-9999.rst:318]: Just as a note: this is only true if the C extension adds new functionality implemented in C (which isn't always the case). It is a subtle point but worth making. If you construct a C type using freezable metaclasses it will itself be freezable.

File: pep-9999.rst [hunk 22]
```diff
@@ -0,0 +1,868 @@
...
+register themselves using ``register_freezable(type)``. To
```
Comment (matajoh) [inline pep-9999.rst:319]: Here is some sample code you could potentially include:

```c
    PyObject *register_freezable = _PyImport_GetModuleAttrString("immutable", "register_freezable");
    if(register_freezable != NULL)
    {
        PyObject* result = PyObject_CallOneArg(register_freezable, (PyObject *)st->Element_Type);
        if(result == NULL){
            goto error;
        }

        Py_DECREF(register_freezable);
    }
```

File: pep-9999.rst [hunk 23]
```diff
@@ -0,0 +1,868 @@
...
+======================
+
+1. Add the ``immutable`` module, the ``NotWriteableError`` type, and
+   the ``NotFreezable`` type.
+2. Add the ``freeze(obj)`` function to the ``immutable`` module and
+   ensure that it traverses object references safely, including
+   cycle detection, and marks objects appropriately, and backs
+   out on failure, without partially freezing object graphs.
+3. Add the ``register_freezable(type)`` whitelisting function.
+4. Add the ``is_freezable(obj)`` function which checks that all
```
Comment (matajoh) [inline pep-9999.rst:619]: isfreezable

File: pep-9999.rst [hunk 24]
```diff
@@ -0,0 +1,868 @@
...
+   objects reachable from ``obj`` can be frozen.
+5. Add the ``register_freezable(type)`` function that is used to
+   whitelist types implemented as C extensions, permitting their
+   instances to be frozen.
+6. Add the ``is_frozen(obj)`` function to the ``immutable`` module
```
Comment (matajoh) [inline pep-9999.rst:624]: isfrozen

File: pep-9999.rst [hunk 25]
```diff
@@ -0,0 +1,868 @@
...
+   that checks whether or not an object is frozen. The status
+   is accessible through ``_Py_ISIMMUTABLE`` in the C API and in
+   Python code through the ``is_frozen(obj)`` function.
+7. Modify object mutation operations (``PyObject_SetAttr``,
+   ``PyDict_SetItem``, ``PyList_SetItem``, etc.) to check the 
+   flag and raise an error when appropriate.
+8. Modify mutation operations in modules in the standard library.
```
Comment (matajoh) [inline pep-9999.rst:631]: Changes to the C ABI:

1. `Py_CHECKWRITE`
2. `_Py_IsImmutable`
3. `PyErr_WriteToImmutable`

Changes to the internal API:
1. `_PyType_HasExtensionSlots(PyTypeObject*)` - Determines whether a TypeObject adds novel functionality in C
2. `_PyNotFreezable_Type`
3. `_PyImmutability_Freeze`
4. `_PyImmutability_RegisterFreezable`
5. `_PyImmutability_IsFreezable`

File: pep-9999.rst [hunk 26]
```diff
@@ -0,0 +1,868 @@
...
+   dictionaries created by its accessors when the object itself is
+   frozen.
+2. The managed buffer type needs to be frozen when the object it
+   is created from is frozen.
+3. Teardown of strongly connected components of immutable objects
+   (see `Simplified Garbage Collection for Immutable Object
+   Graphs`_) must be hooked into ``Py_DECREF``.
+
+As such, we implement a function which is not in the limited API
+(and thus not part of the stable C ABI) called ``_Py_Freeze``
```
Comment (matajoh) [inline pep-9999.rst:697]: Note the name change above

**Review by matajoh — COMMENTED** [review]
I've gone through and made some updates based on implementation details

File: pep-9999.rst (hunk 19 above)
Comment (TobiasWrigstad) [reply pep-9999.rst:226]: I added it in but dropped the word "implicitly" because that was confusing.

File: pep-9999.rst (hunk 25 above)
Comment (TobiasWrigstad) [reply pep-9999.rst:631]: I'll add these as subsections under the list. Sounds OK?

File: pep-9999.rst [hunk 27]
```diff
@@ -0,0 +1,937 @@
...
+   which must either be ``True`` or ``False``. In the first case,
+   ``freeze`` behaves as in alternative 1), in the second case,
+   it behaves as in alternative 2). We could further track whether
+   an object is strictly immutable or not in order to prevent
+   non-strictly immutable objects to participate in operations
+   which require strictness. This adds additional complexity to
+   the implementation, and also for the user.
+   
+This PEP proposes following alternative 1, where freezing either
+succeeds or fails completely.
```
Comment (xFrednet) [inline pep-9999.rst:306]: Shouldn't this say: "or fails partially"? Option 2 would be failing completely.

```suggestion
This PEP proposes following alternative 1, where freezing either
succeeds or fails partially.
```

Or it could be that the number is wrong and this pep actually proposes success or failing completely which would be "2. rejected alternative"

---

I also wonder if we should add a motivation like "[...] fails partially. We believe this is easier to implement and more in the spirit of what the user wants."

But this could, of course, spark discussions

File: pep-9999.rst [hunk 28]
```diff
@@ -0,0 +1,937 @@
...
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Due to the opt-in decision, there are no *obligations* for C
+extensions that does not want to add support for immutability.
```
Comment (xFrednet) [inline pep-9999.rst:313]:
```suggestion
extensions that do not want to add support for immutability. 
```

Spelling error?

File: pep-9999.rst [hunk 29]
```diff
@@ -0,0 +1,937 @@
...
+  classes. This means that any metaprogramming or changes to a class
+  must happen before a class is frozen. **Mitigation:** A frozen class
+  can be extended and its behaviour overridden through normal object-oriented
+  means. If neccessary, it is possible to add an option to make a mutable
+  copy of immutable objects and classes, which could then be changed.
+  Mutable instances of an immutable class can have their classes changed
+  to the mutable copy by reassigning ``__class__``.
+* Metaclass Freezing: Since class objects have metaclasses,
+  freezing a class may propagate upwards through the metaclass
+  hierarchy. This means that the ``Type`` object will be made immutable
```
Comment (xFrednet) [inline pep-9999.rst:469]: Above it was called `type` with a lower letter, which, I believe, is the correct way?

```suggestion
  hierarchy. This means that the ``type`` object will be made immutable
```

File: pep-9999.rst [hunk 30]
```diff
@@ -0,0 +1,937 @@
...
+
+Implementation Details
+======================
+
+1. Add the ``immutable`` module, the ``NotWriteableError`` type, and
+   the ``NotFreezable`` type.
+2. Add the ``freeze(obj)`` function to the ``immutable`` module and
+   ensure that it traverses object references safely, including
+   cycle detection, and marks objects appropriately, and backs
+   out on failure, without partially freezing object graphs.
```
Comment (xFrednet) [inline pep-9999.rst:637]: Doesn't section "Dealing with Failure During Freezing" suggest taking the first option with partial freezing? :thinking:

File: pep-9999.rst [hunk 31]
```diff
@@ -0,0 +1,937 @@
...
+3. Add the ``register_freezable(type)`` whitelisting function.
+4. Add the ``isfreezable(obj)`` function which checks that all
+   objects reachable from ``obj`` can be frozen.
+5. Add the ``register_freezable(type)`` function that is used to
+   whitelist types implemented as C extensions, permitting their
+   instances to be frozen.
```
Comment (xFrednet) [inline pep-9999.rst:643]: Are 3. and 5. supposed to have the same function name? Should we remove 3.

**Review by xFrednet — COMMENTED** [review]
Very good changes, I mainly have some nits and thinks I was unsure about when reading

File: pep-9999.rst (hunk 31 above)
Comment (TobiasWrigstad) [reply pep-9999.rst:643]:
```suggestion
3. Add the ``isfreezable(obj)`` function which checks that all
   objects reachable from ``obj`` can be frozen.
4. Add the ``register_freezable(type)`` function that is used to
   whitelist types implemented as C extensions, permitting their
   instances to be frozen.
```

File: pep-9999.rst (hunk 30 above)
Comment (TobiasWrigstad) [reply pep-9999.rst:637]:
```suggestion
2. Add the ``freeze(obj)`` function to the ``immutable`` module and
   ensure that it traverses object references safely, including
   cycle detection, and marks objects appropriately, and backs
   out on failure, possibly partially freezing the object graph.
```
